osu.Game/Online/HubClientConnector.cs

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable enable
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR.Client;
+using Microsoft.Extensions.DependencyInjection;
+using Newtonsoft.Json;
+using osu.Framework;
+using osu.Framework.Bindables;
+using osu.Framework.Logging;
+using osu.Game.Online.API;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// A component that maintains over a hub connection between client and server.
+    /// </summary>
+    public class HubClientConnector : IDisposable
+    {
+        /// <summary>
+        /// Invoked whenever a new hub connection is built.
+        /// </summary>
+        public Action<HubConnection>? OnNewConnection;
```

Comment:
This should probably be renamed to `ConfigureConnection` or similar. Right now it's not clear whether this is fired before or after a connection occurs (the method name suggests after).